File: guests/python/feed.py
"""The feed scene, sum-typed elements (tools/scenes/feed.steps).

    KAYA_SELFTEST=feed python3 guests/python/feed.py
"""

import sys
from dataclasses import dataclass

import kaya


@dataclass
class Note:
    text: str


@dataclass
class Todo:
    title: str
    done: bool


app = kaya.App()


def done_count_text(items):
    n = sum(1 for p in items.values() if isinstance(p, Todo) and p.done)
    return f"{n} done"


def on_promote():
    # The MODEL says which entry is a Note; the update restamps in place.
    for key, post in feed.items():
        if isinstance(post, Note):
            feed.update(key, Todo(title=post.text, done=True))
            break


def on_toggle(key, checked):
    # A stale occurrence lands in the else and folds into nothing.
    if isinstance(feed.get(key), Todo):
        feed.patch(key, done=checked)


with app.window():
    feed = kaya.collection(Note | Todo)
    done_count = feed.derive(done_count_text)
    with kaya.row():
        kaya.button("promote", on_click=on_promote)
        kaya.label(bind=done_count)
        with kaya.for_each(feed) as cases:
            with cases.case(Note) as note:
                kaya.label(bind=note.text)
            with cases.case(Todo) as todo:
                with kaya.row():
                    kaya.checkbox(checked=todo.done, on_toggle=on_toggle)
                    kaya.label(bind=todo.title)
    feed.insert("a", Note(text="jot one"))
    feed.insert("b", Todo(title="buy milk", done=False))
    feed.insert("c", Note(text="jot two"))

sys.exit(app.run())

File: guests/js/feed.ts
// The list of record types IS the sum, in that order, and the eliminator
// is held to totality at declaration. See guests/rust/feed.rs.
//     KAYA_SELFTEST=feed node guests/js/feed.ts

import * as kaya from "kaya-gui";

const Note = kaya.record({ text: String }, "Note");
const Todo = kaya.record({ title: String, done: Boolean }, "Todo");
type Post = kaya.Fields<typeof Note.schema> | kaya.Fields<typeof Todo.schema>;

const app = new kaya.App();

function doneCountText(items: Map<kaya.Key, Post>): string {
  let n = 0;
  for (const p of items.values()) if (p instanceof Todo && p.done) n += 1;
  return `${n} done`;
}

function onPromote(): void {
  // The MODEL says which entry is a Note; the update restamps in place.
  for (const [key, post] of feed.items()) {
    if (post instanceof Note) {
      feed.update(key, Todo({ title: post.text, done: true }));
      break;
    }
  }
}

function onToggle(post: kaya.RowHandle<Post>, checked: boolean): void {
  // A row that has left matches no variant: a stale occurrence folds away.
  if (post instanceof Todo) post.done = checked;
}

let feed!: kaya.Collection<Post, kaya.Cases>;

app.window(() => {
  feed = kaya.collection([Note, Todo]);
  const doneCount = feed.derive(doneCountText);
  kaya.row(() => {
    kaya.button("promote", { onClick: onPromote });
    kaya.label({ bind: doneCount });
    kaya.forEach(feed, (cases) => {
      cases.case(Note, (note) => {
        kaya.label({ bind: note.text });
      });
      cases.case(Todo, (todo) => {
        kaya.row(() => {
          kaya.checkbox({ checked: todo.done, onToggle });
          kaya.label({ bind: todo.title });
        });
      });
    });
  });
  feed.insert("a", Note({ text: "jot one" }));
  feed.insert("b", Todo({ title: "buy milk", done: false }));
  feed.insert("c", Note({ text: "jot two" }));
});

app.run();
